RCNN_tf.py

```diff
@@ -104,97 +81,80 @@ def rcl(input_data, num_input_chans, num_filter, filter_shape, num_of_data,
-
-        # TODO I am not sure why but this learns quicker. Investigate?
-        cond = termination_cond()
-        results = tf.while_loop(
-            cond=cond,
-            body=loop_body,
-            loop_vars=[0, cell_states, output],
-        )
-
-        output = results[2]
-
-        return output
+    output = forward_output
+    output = tf.nn.bias_add(output, bias)
+    cell_states = tf.nn.local_response_normalization(
+        tf.maximum(np.asarray(0.0).astype(PRECISION_NP), output),
+        depth_radius=num_norm_feat_maps,
+        bias=1,
+        alpha=alpha,
+        beta=beta,
+        name=name + '_lrn'
+    )
 
-    elif "GPU" in processing_unit.upper():
-        # Use this loop instead of the tensorflow while loop since it causes
-        # trouble running it on the GPU.  When it is used without the optimiser
-        # it is assumed to be never executed, and hence, a dead end.
-        for _ in range(depth):
-            recurrent_output = tf.nn.conv2d(
-                cell_states,
+    def termination_cond():
+        """Specifies the termination condition for the Tensorflow
+        while loop"""
+        return lambda x, recurrent_states, output: x < depth
+
+    def loop_body(x, recurrent_states, output):
+        """Specifies the body callable of the Tensorflow while loop
+
+        Args/Returns:
+            x                (tf.Tensor): counter
+            recurrent_states (tf.Tensor): the recurrent states of the
+                                          recurrent convolutional layer
+            output           (tf.Tensor): the output tensor, which is the
+                                          addition of the forward and
+                                          recurrent output and the bias
+        """
+        recurrent_output = tf.nn.conv2d(
+                recurrent_states,
                 recurrent_weights,
                 [1, 1, 1, 1],
-                padding='SAME'
-            )
+                padding='SAME')
+
+        output = tf.add(forward_output, recurrent_output)
+        output = tf.nn.bias_add(output, bias)
+
+        recurrent_states = tf.nn.local_response_normalization(
+            tf.maximum(np.asarray(0.0).astype(PRECISION_NP), output),
+            depth_radius=num_norm_feat_maps,
+            bias=1,
+            alpha=alpha,
+            beta=beta,
+            name=name + '_lrn'
+        )
 
-            output = tf.add(forward_output, recurrent_output)
-            output += bias
+        x += 1
 
-            cell_states = tf.nn.local_response_normalization(
-                tf.maximum(np.asarray(0.0).astype(PRECISION_NP), output),
-                depth_radius=num_norm_feat_maps,
-                bias=1,
-                alpha=alpha,
-                beta=beta,
-                name=name + '_lrn'
-            )
+        return x, recurrent_states, output
 
-        return output
+    # TODO I am not sure why but this learns quicker. Investigate?
```

Comment:
It probably has to do with the order in which Tensorflow calls the functions in the while loop.
We could investigate it, but if you have seen that it is faster I think we could just leave it as it is.

Reply:
Yea think so, too. For the moment we can leave it as it is